File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/integration/type/nbt/NbtModificationResult.java
package io.github.arcaneplugins.levelledmobs.bukkit.integration.type.nbt;

import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import org.bukkit.entity.Entity;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

public class NbtModificationResult {

    private ItemStack itemStack;
    private Entity entity;
    private Exception exception;
    private final List<String> objectsAdded = new LinkedList<>();
    private final List<String> objectsUpdated = new LinkedList<>();
    private final List<String> objectsRemoved = new LinkedList<>();

    public NbtModificationResult(final ItemStack itemStack) {
        this.itemStack = itemStack;
    }

    public NbtModificationResult(final Entity entity) {
        this.entity = entity;
    }

    @NotNull
    public NbtModificationResult withException(final Exception exception) {
        this.exception = exception;
        return this;
    }

    @NotNull
    public NbtModificationResult withObjectsAdded(final Object... objects) {
        for(var object : objects)
            getObjectsAdded().add(object.toString());
        return this;
    }

    @NotNull
    public NbtModificationResult withObjectsUpdated(final Object... objects) {
        for(var object : objects)
            getObjectsUpdated().add(object.toString());
        return this;
    }

    @NotNull
    public NbtModificationResult withObjectsRemoved(final Object... objects) {
        for(var object : objects)
            getObjectsRemoved().add(object.toString());
        return this;
    }

    @NotNull
    public Entity getEntity() {
        return Objects.requireNonNull(entity);
    }

    public boolean hasEntity() {
        return entity != null;
    }

    @NotNull
    public ItemStack getItemStack() {
        return Objects.requireNonNull(itemStack);
    }

    public boolean hasItemStack() {
        return itemStack != null;
    }

    @NotNull
    public Exception getException() {
        return Objects.requireNonNull(exception);
    }

    public boolean hasException() {
        return exception != null;
    }

    @NotNull
    public List<String> getObjectsAdded() {
        return objectsAdded;
    }

    @NotNull
    public List<String> getObjectsUpdated() {
        return objectsUpdated;
    }

    @NotNull
    public List<String> getObjectsRemoved() {
        return objectsRemoved;
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/integration/type/nbt/NbtModificationResult.kt
package io.github.arcaneplugins.levelledmobs.bukkit.integration.type.nbt

import org.bukkit.entity.Entity
import org.bukkit.inventory.ItemStack
import java.lang.Exception

class NbtModificationResult() {
    constructor(itemStack: ItemStack) : this(){
        this.itemStack = itemStack
    }

    constructor(entity: Entity) : this(){
        this.entity = entity
    }

    var itemStack: ItemStack? = null
        private set
    var entity: Entity? = null
        private set
    var exception: Exception? = null
        private set
    private val objectsAdded: MutableList<String> = mutableListOf()
    private val objectsUpdated: MutableList<String> = mutableListOf()
    private val objectsRemoved: MutableList<String> = mutableListOf()

    fun withException(exception: Exception):  NbtModificationResult{
        this.exception = exception
        return this
    }

    fun withObjectsAdded(objects: MutableList<Any>): NbtModificationResult{
        for (obj in objects){
            objectsAdded.add(obj.toString())
        }
        return this
    }

    fun withObjectsUpdated(objects: MutableList<Any>): NbtModificationResult{
        for (obj in objects){
            objectsUpdated.add(obj.toString())
        }
        return this
    }

    fun withObjectsRemoved(objects: MutableList<Any>): NbtModificationResult{
        for (obj in objects){
            objectsRemoved.add(obj.toString())
        }
        return this
    }

    val hasEntity: Boolean
        get() {
            return this.entity != null
        }

    val hasItemStack: Boolean
        get() {
            return this.itemStack != null
        }

    val hasException: Boolean
        get() {
            return this.exception != null
        }
}